roles/openshift_health_checker/openshift_checks/docker_image_availability.py

```diff
@@ -0,0 +1,147 @@
+
+    This check attempts to pull required docker images and fail if
+    any image is missing or an error occurs during pull.
+    """
+
+    name = "docker_image_availability"
+    tags = ["preflight"]
+
+    docker_path = "/bin/docker"
+    skopeo_image = "juanvallejo/skopeo"
+
+    def run(self, tmp, task_vars):
+        is_containerized = get_var(task_vars, "openshift", "common", "is_containerized")
+        openshift_release = get_var(task_vars, "openshift_release")
+        openshift_image_tag = get_var(task_vars, "openshift_image_tag")
+        deployment_type = get_var(task_vars, "deployment_type")
+
+        registry_prefix = "openshift3/ose"
+        if deployment_type == "origin":
+            registry_prefix = "openshift/origin"
+
+        required_images = set(self.rpm_docker_images(openshift_release))
+        if is_containerized:
+            required_images = set(self.containerized_docker_images(registry_prefix, openshift_release))
+
+        # append images with qualified image tags to our list of required images.
+        # these are images with a (v0.0.0.0) tag, rather than a standard release
+        # format tag (v0.0). We want to check this set in both containerized and
+        # non-containerized installations.
+        required_images.update(self.qualified_docker_images(registry_prefix, "v" + openshift_image_tag))
+
+        # use docker_image_facts to inspect local images before attempting to use Skopeo
+        local_images = self.inspect_local_images(required_images, task_vars)
+        required_images = local_images.get("failed_images")
+
+        # ensure skopeo image exists or can be pulled and used successfully
+        skopeo_image = self.check_skopeo_image(self.docker_path, self.skopeo_image, tmp, task_vars)
+        if "failed" in skopeo_image:
+            msg = "Failed to pull or use existing Skopeo image: %s" % skopeo_image.get("msg")
+            return {"failed": True, "msg": msg}
+
+        docker_info_module = self.module_executor("docker_info", {}, task_vars)
+
+        # use docker info to determine which registries
+        # Skopeo should inspect for required images
+        docker_info = docker_info_module.get("info")
+        if not docker_info:
+            return {"failed": True, "msg": "Unable to get docker host info. Is `docker-py` installed?"}
+
+        docker_regs = docker_info.get("Registries")
+        if not docker_regs:
+            return {"failed": True, "msg": "Unable to read docker registry data from host."}
+
+        regs = [
+            reg.get("Name") for reg in docker_regs
+        ]
+
+        # inspect all rpm or containerized docker images for every docker registry found
+        while len(regs) and len(required_images):
+            current_reg = regs[0]
+            inspect = self.inspect_images(self.docker_path, self.skopeo_image, current_reg, required_images, tmp, task_vars)
+            required_images = inspect.get("failed_images")
```

Comment:
This block is exponential time at worst, however a successful local check helps the number of images it has to process / skips this completely. Also, if skopeo is not able to be pulled, or doesn't exist ahead of time locally, or the "skopeo image" does not contain the `skopeo` bin for whatever reason, the check fails early before getting to this block. Thoughts?